### ISSUE TITLE
Rule Enhancement Proposal: func-names with never

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6059]
**EDIT:** this proposal was updated in https://github.com/eslint/eslint/issues/6059#issuecomment-216719718

----

From [requireAnonymousFunctions] and [requireMatchingFunctionName].

This proposal is to add some options.

```js
{
    "func-names": [
        "error",
        "always" or "never",
        {
            "sameAsVariable": false
        }
}
```

- `"always"` (default) - This is the current behavior.
- `"never"` - This will disallow a use of function expression names.
- `sameAsVariable` (default is `false`) - If this is `true`, the function expression name must be the same as the variable name which is the destination of the assignment. For example: `let foo = function foo() {};`. But if the destination is `module.exports`, this rule will ignore it.

[requireAnonymousFunctions]: http://jscs.info/rule/requireAnonymousFunctions
[requireMatchingFunctionName]: http://jscs.info/rule/requireMatchingFunctionName

**Comment by platinumazure** [comment]
Should `sameAsVariable: true` also check for assignment to a MemberExpression at all? (e.g., `foo.bar = function baz () { }`)

Should we allow a "tolerate" value for the first option, so that a user can decide that s/he does not always want function names, but when a function name is specified, it must match the assigned variable name?

**Comment by mysticatea** [comment]
> Should sameAsVariable: true also check for assignment to a MemberExpression at all? (e.g., foo.bar = function baz () { })

Exactly.

> Should we allow a "tolerate" value for the first option, so that a user can decide that s/he does not always want function names, but when a function name is specified, it must match the assigned variable name?

Ah, interesting.
Maybe `requireMatchingFunctionName` should not merge with `func-names` rule.

**Comment by platinumazure** [comment]
> > Should sameAsVariable: true also check for assignment to a MemberExpression at all? (e.g., foo.bar = function baz () { })
> 
> Yes.

So if `sameAsVariable: true`, does this mean users cannot assign functions to object members at all?

> Ah, interesting.  
> Maybe requireMatchingFunctionName should not merge with func-names rule.

I think I agree-- that could be a separate rule, which wouldn't do much if function expressions don't have names, but it would execute safely and not confuse the options schema of this proposed rule.

**Comment by mysticatea** [comment]
I update this proposal:

----

From [requireAnonymousFunctions].

This proposal is to add `"never"` option to `func-names` rule.

```js
{
    "func-names": [
        "error",
        "always" or "never"
    ]
}
```

- `"always"` (default) - This is the current behavior.
- `"never"` - This will disallow a use of function expression names.

**Valid:**

```js
/*eslint func-names: ["error", "never"]*/

function foo() {
}
var foo = function() {
};
```

**Invalid:**

```js
/*eslint func-names: ["error", "never"]*/

var foo = function foo() {
};
```

[requireAnonymousFunctions]: http://jscs.info/rule/requireAnonymousFunctions

**Comment by Mouvedia** [comment]
related: #938

**Comment by alberto** [comment]
@eslint/eslint-team opinions? 

I can champion this one

**Comment by mikesherov** [comment]
:+1:

**Comment by Mouvedia** [comment]
👍 

Here's another reason—if you care about cross-browser issues—to make this happen: https://kangax.github.io/nfe/#safari-bug

**Comment by nzakas** [comment]
Looks like we are good to move forward here. 

@Mouvedia this issue is for creating an option that is compatible with JSCS functionality. If you want more than that, please open a new issue.

**Comment by Mouvedia** [comment]
@nzakas after 3 refusals, I don't think Ill submit a new one…
